envoy/common/io/io_uring.h

```diff
@@ -12,8 +11,15 @@ namespace Io {
  * @param user_data is any data attached to an entry submitted to the submission
  * queue.
  * @param result is a return code of submitted system call.
+ * @param injected indicates whether the completion is injected or not.
+ */
+using CompletionCb = std::function<void(void* user_data, int32_t result, bool injected)>;
```

Comment:
Right now C++ classes closely resemble the C io_uring API. Is it possible to make it more C++ and avoid the use of the `void*`. Conceptually io_uring allows submitting I/O operation and the completion callback. Would it make more sense for the `prepare*` calls to get the completion callback instead of the `void*` and then hide all of the `void*` business that C API needs inside the IoUring class?

Reply:
That `user_data` is the data related to each request. I add C++ style `BaseRequest` to carry those user data. https://github.com/envoyproxy/envoy/pull/24958/files#diff-df9a08b7c36ffbd691c64a9c0c22a34cd2848af1ce434bc05bd661dbba4b7199R16

There are also a set of subclasses, like AcceptRequest, ReadRequest...etc 

The class `IoUring` is a pretty simple wrap of iouring C interface and low-level interface. so it still exposes that C-style stuff. I'm planning to hide all those details in the IoUringSocket and IoUringWorker interface. 
https://github.com/envoyproxy/envoy/pull/24958/files#diff-dad289fe5a60a8403503d6ec8e79baeb72f4de7a40aaba36646494cdfde1fd39R199
https://github.com/envoyproxy/envoy/pull/24958/files#diff-dad289fe5a60a8403503d6ec8e79baeb72f4de7a40aaba36646494cdfde1fd39R344

And I suppose only IoUringWorker will use the IoUring interface. All the up-layer code of Envoy only uses the IoUringWorker and IoUringSocket interface.

For the user of Iouring `IoUringSocketHandleImpl` won't see any `void*` and those C-style interfaces since it only uses the IoUringWorker and IoUringSocket.

Does that make sense?

Reply:
Or I can change the interface like this 
`using CompletionCb = std::function<void(BaseRequest* user_data, int32_t result, bool injected)>;`

And prepare* also accept the `BaseRequest *`

Reply:
Yes, the change you are proposing is good.

Reply:
After thinking, let me change this in this PR, then I needn't the user data releasor anymore.